### PR TITLE
Item config & level up logic

### PR DESCRIPTION
**Pull request by lotuuu**
Implements ItemTemplate configuration and tier up (item fusion) functionalities.

Item tier ups work by combining many items into a new one. Unlike units, where we delete all units but one and raise the latter's tier, here we create a new item altogether. Reason is we don't have any more data other than the unit it's equipped to: The template defines the item now, they don't hold a level themselves. So really there's not really an issue with removing all items entirely.

Source of truth: https://docs.google.com/spreadsheets/d/177mvJS75LecaAEpyYotQEcrmhGJWI424UnkE2JHLmyY/edit#gid=1542073627

# To Review
Most of the change lines come from the item_templates.json file. The main file changes on this PR are:
- `apps/champions/lib/champions/config.ex`, where we use some new utility functions defined in `GameBackend.Items` to import the item template config
- `apps/champions/lib/champions/items.ex`, where we define the logic for item fusion
- `apps/game_backend/lib/game_backend/items/item_template.ex` and `apps/game_backend/priv/repo/migrations/20240425171046_add_item_template_fields.exs`, where we define the new fields for the ItemTemplate schema.

Of course, this is just meant to help make PR review process a bit easier given its large amount of files and lines changed, but do check all the of them anyways!

# To test

You can test item fusion by running `Champions.Items.fuse(user_id, [item_1.id, item_2.id, item_3.id])`. Remember that when you define your items' template, you also need to define the template for what they will upgrade into! Check the tests added [here](https://github.com/lambdaclass/mirra_backend/blob/d796ffe9b1c0a5d25a2ed88e6d6f1a8a0b2a5fbc/apps/gateway/test/champions_test.exs#L731) for an example.

### CODE REVIEW

**Review by ncontinanza — CHANGES_REQUESTED** [review]

**Review by ncontinanza — APPROVED** [review]

File: apps/champions/lib/champions/items.ex [hunk 1]
```diff
@@ -12,36 +14,94 @@ defmodule Champions.Items do
...
+          Currencies.substract_currencies(user_id, new_template.upgrade_costs)
+        end)
+        |> Transaction.run()
+
+      case result do
+        {:error, _, _, _} ->
+          {:error, :transaction}
+
+        {:ok, %{item: item}} ->
+          Items.get_item(item.id)
```
Comment (agustinesco) [inline apps/champions/lib/champions/items.ex:68]: why not simply returning the item?

File: apps/champions/lib/champions/items.ex [hunk 2]
```diff
@@ -12,36 +14,94 @@ defmodule Champions.Items do
...
+      end
+    else
+      {:consumed_items_owned, false} ->
+        {:error, :consumed_items_not_found}
```
Comment (agustinesco) [inline apps/champions/lib/champions/items.ex:72]:
```suggestion
        {:error, :consumed_items_not_owned}
```

File: apps/champions/lib/champions/items.ex [hunk 3]
```diff
@@ -12,36 +14,94 @@ defmodule Champions.Items do
...
+
+      {:consumed_items_valid, {false, _}} ->
+        {:error, :consumed_items_invalid}
+    end
+  end
+
+  defp delete_consumed_items(item_ids) do
+    {amount_deleted, _return} = Items.delete_items(item_ids)
+
+    if Enum.count(item_ids) == amount_deleted, do: {:ok, amount_deleted}, else: {:error, "failed"}
```
Comment (agustinesco) [inline apps/champions/lib/champions/items.ex:85]: In which situations this would fail?

File: apps/champions/lib/champions/items.ex (hunk 2 above)
Comment (lotuuu) [reply apps/champions/lib/champions/items.ex:72]: I chose to hide the fact that an item with the given id actually exists when the user doesn't own it. It's none of their business!

File: apps/champions/lib/champions/items.ex (hunk 1 above)
Comment (lotuuu) [reply apps/champions/lib/champions/items.ex:68]: This is to get the preloads that protobuf messages need

**Review by tkz00 — COMMENTED** [review]
Do you think adding `Champions.Items.fuse` function to the readme file would be a good idea?

**Review by tkz00 — APPROVED** [review]

File: apps/champions/lib/champions/items.ex (hunk 2 above)
Comment (agustinesco) [reply apps/champions/lib/champions/items.ex:72]: right!

**Review by agustinesco — APPROVED** [review]

**Pull request merged**